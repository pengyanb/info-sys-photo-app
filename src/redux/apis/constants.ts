export const ACTION_GET_ALBUMS = "ACTION_GET_ALBUMS";
export const ACTION_GET_ALBUMS_RESPONSE = "ACTION_GET_ALBUMS_RESPONSE";
export const ACTION_GET_ALBUMS_FAILED = "ACTION_GET_ALBUMS_FAILED";

export const ACTION_GET_PHOTOS = "ACTION_GET_PHOTOS";
export const ACTION_GET_PHOTOS_RESPONSE = "ACTION_GET_PHOTOS_RESPONSE";
export const ACTION_GET_PHOTOS_FAILED = "ACTION_GET_PHOTOS_FAILED";