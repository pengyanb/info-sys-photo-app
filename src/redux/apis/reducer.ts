import {
  ACTION_GET_ALBUMS,
  ACTION_GET_ALBUMS_RESPONSE,
  ACTION_GET_ALBUMS_FAILED,
  ACTION_GET_PHOTOS,
  ACTION_GET_PHOTOS_RESPONSE,
  ACTION_GET_PHOTOS_FAILED
} from "./constants";

import {
  IAlbumInfo,
  IPhotoInfo,
  IActionGetAlbums,
  IActionGetAlbumsResponse,
  IActionGetAlbumsFailed,
  IActionGetPhotos,
  IActionGetPhotosResponse,
  IActionGetPhotosFailed
 } from "./actions";

 export interface IApiState {
  albums: IAlbumInfo[];
  albumsError: any;
  photos: IPhotoInfo[];
  photosError: any
 } 

const initialState: IApiState = {
  albums: [],
  albumsError: undefined,
  photos: [],
  photosError: undefined
};

const apiReducer = (state: IApiState = initialState, action: IActionGetAlbums | IActionGetAlbumsResponse | IActionGetAlbumsFailed | IActionGetPhotos | IActionGetPhotosResponse | IActionGetPhotosFailed) => {
  switch (action.type) {
    case ACTION_GET_ALBUMS:
      return {
        ...state,
        albums: [],
        albumsError: undefined,
      };
    case ACTION_GET_ALBUMS_RESPONSE:
      return {
        ...state,
        albums: (action as IActionGetAlbumsResponse).albums,
        albumsError: undefined,
      };
    case ACTION_GET_ALBUMS_FAILED:
      return {
        ...state,
        albums: [],
        albumsError: (action as IActionGetAlbumsFailed).error,
      };
    case ACTION_GET_PHOTOS:
      return {
        ...state,
        photos:[],
        photosError: undefined,
      };
    case ACTION_GET_PHOTOS_RESPONSE:
      return {
        ...state,
        photos: (action as IActionGetPhotosResponse).photos,
        photosError: undefined
      };
    case ACTION_GET_PHOTOS_FAILED:
      return {
        ...state,
        photos: [],
        photosError: (action as IActionGetAlbumsFailed).error
      };
    default:
      return state;
  }
};

export default apiReducer;