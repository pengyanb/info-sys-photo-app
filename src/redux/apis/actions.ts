import { ACTION_GET_ALBUMS, ACTION_GET_ALBUMS_RESPONSE, ACTION_GET_ALBUMS_FAILED, ACTION_GET_PHOTOS, ACTION_GET_PHOTOS_RESPONSE, ACTION_GET_PHOTOS_FAILED } from "./constants";

export interface IAlbumInfo {
  userId: number;
  id: number;
  title: string;
}

export interface IPhotoInfo {
  albumId: number;
  id: number;
  title: string;
  url: string;
  thumbnailUrl: string;
}

export interface IActionGetAlbums {
  type: string;
}

export interface IActionGetAlbumsResponse {
  type: string;
  albums: IAlbumInfo[];
}

export interface IActionGetAlbumsFailed {
  type: string;
  error: any;
}

export interface IActionGetPhotos {
  type: string;
}

export interface IActionGetPhotosResponse {
  type: string;
  photos: IPhotoInfo[];
}

export interface IActionGetPhotosFailed {
  type: string;
  error: any;
}

export function actionGetAlbums(): IActionGetAlbums {
  return {
    type: ACTION_GET_ALBUMS
  }
}

export function actionGetAlbumsResponse(albums: IAlbumInfo[]): IActionGetAlbumsResponse {
  return {
    type: ACTION_GET_ALBUMS_RESPONSE,
    albums
  };
}

export function actionGetAlbumsFailed(error: any): IActionGetAlbumsFailed {
  return {
    type: ACTION_GET_ALBUMS_FAILED,
    error
  }
}

export function actionGetPhotos(): IActionGetPhotos {
  return {
    type: ACTION_GET_PHOTOS
  }
}

export function actionGetPhotosResponse(photos: IPhotoInfo[]): IActionGetPhotosResponse {
  return {
    type: ACTION_GET_PHOTOS_RESPONSE,
    photos
  }
}

export function actionGetPhotosFailed(error: any): IActionGetPhotosFailed {
  return {
    type: ACTION_GET_PHOTOS_FAILED,
    error
  }
}