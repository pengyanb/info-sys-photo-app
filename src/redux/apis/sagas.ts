import { call, put, takeLatest, all } from "redux-saga/effects"
import axios from "axios";
import { ACTION_GET_ALBUMS, ACTION_GET_PHOTOS } from "./constants";
import { actionGetAlbumsResponse, actionGetAlbumsFailed, actionGetPhotosResponse, actionGetPhotosFailed } from "./actions";

const CONST_ALBUMS_URL = "http://jsonplaceholder.typicode.com/albums";
const CONST_PHOTOS_URL = "http://jsonplaceholder.typicode.com/photos";

export function* getAlbumsRequest() {
  
  try {
    const result = yield call((axios as any), {
      url: CONST_ALBUMS_URL,
      method: "GET"
    });
    console.log("!!!!getAlbumsRequest", result.data);
    yield put(actionGetAlbumsResponse(result.data));
  } catch (err) {
    yield put(actionGetAlbumsFailed(err));
  }
};

export function* getPhotosRequest() {
  try {
    const result = yield call((axios as any), {
      url: CONST_PHOTOS_URL,
      method: "GET"
    });
    yield put(actionGetPhotosResponse(result.data));
  } catch (err) {
    yield put(actionGetPhotosFailed(err));
  }
}

export function* getAlbumsWatcher() {
  yield takeLatest(ACTION_GET_ALBUMS, getAlbumsRequest);
}

export function* getPhotosWatcher() {
  yield takeLatest(ACTION_GET_PHOTOS, getPhotosRequest);
}

export default function* apiSaga() {
  yield all([getAlbumsWatcher(), getPhotosWatcher()]);
}