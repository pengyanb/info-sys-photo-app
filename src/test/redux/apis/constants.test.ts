import * as Constants from "../../../redux/apis/constants";

describe("redux/apis/constants tests", () => {
  it("constants snapshot test", () => {
    expect(Constants).toMatchSnapshot();
  });
});