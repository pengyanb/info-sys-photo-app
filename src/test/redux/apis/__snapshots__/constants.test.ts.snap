// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/apis/constants tests constants snapshot test 1`] = `
Object {
  "ACTION_GET_ALBUMS": "ACTION_GET_ALBUMS",
  "ACTION_GET_ALBUMS_FAILED": "ACTION_GET_ALBUMS_FAILED",
  "ACTION_GET_ALBUMS_RESPONSE": "ACTION_GET_ALBUMS_RESPONSE",
  "ACTION_GET_PHOTOS": "ACTION_GET_PHOTOS",
  "ACTION_GET_PHOTOS_FAILED": "ACTION_GET_PHOTOS_FAILED",
  "ACTION_GET_PHOTOS_RESPONSE": "ACTION_GET_PHOTOS_RESPONSE",
}
`;
